Source/WebCore/dom/Document.cpp

```diff
@@ -8933,6 +8933,66 @@ HTMLDialogElement* Document::activeModalDialog() const
     return nullptr;
 }
 
+HTMLElement* Document::topmostAutoPopover() const
+{
+    for (auto& element : makeReversedRange(m_topLayerElements)) {
+        if (auto* candidate = downcast<HTMLElement>(element.ptr()); candidate->popoverState() == PopoverState::Auto)
+            return candidate;
+    }
+
+    return nullptr;
+}
+
+// https://html.spec.whatwg.org/#hide-all-popovers-until
+void Document::hideAllPopoversUntil(Element* endpoint, FocusPreviousElement focusPreviousElement, FireEvents fireEvents)
+{
+    Vector<Ref<HTMLElement>> popoversToHide;
+    for (auto& element : makeReversedRange(m_topLayerElements)) {
+        if (!is<HTMLElement>(element) || downcast<HTMLElement>(element.get()).popoverState() != PopoverState::Auto)
+            continue;
+        if (element.ptr() == endpoint)
+            break;
+        popoversToHide.append(downcast<HTMLElement>(element.get()));
+    }
+
+    for (auto& popover : popoversToHide)
+        popover->hidePopoverInternal(focusPreviousElement, fireEvents);
+}
+
+void Document::handlePopoverLightDismiss(Event& event)
+{
+    ASSERT(event.isTrusted());
+
+    RefPtr topmostAutoPopover = this->topmostAutoPopover();
+    if (!topmostAutoPopover)
+        return;
+
+    RefPtr popoverToAvoidHiding = [&]() -> HTMLElement* {
+        for (Node* node = downcast<Node>(event.target()); node; node = node->parentInComposedTree()) {
```

Comment:
Can we something more like `lineageOfType<HTMLElement>` here?

Reply:
The only thing preventing from using `lineageOfType` is really the fact the `event.target()` is a `Node`, not an `Element`.

Reply:
I thought that function takes a `Node` argument. I remember Antti did it that way.

Reply:
I see now, `ancestorsOfType` takes `Node`, but `lineageOfType` takes `Element`. This might be a design mistake. Should talk to Antti about it, so we have the best pattern for long-term use.

Reply:
I think it might have been intentional, since `lineageOfType` includes itself which has to be an `Element` type in order for `lineageOfType<ElementType>()` to make sense (you can't do `ancestorsOfType<Node>()` or `lineageOfType<Node>()`.

But I can check with the relevant people to see if this can be improved in some way.

Reply:
Yes, of course it was intentional as far as that goes. But if what you have is a node that _might_ be an element, forcing you to do the tape check explicitly, if it might be an element of the type that `lineage` is looking for, may not be the best design.